### DISCUSSION TITLE
Is there abstracted support for file uploads?

### DISCUSSION THREAD

**Discussion by forthrin** [HoneyryderChuck/httpx#37] [Q&A]
Does the library have helpers to generate MIME fluff for file uploads, or must one make them oneself?

Couldn't find any documentation or issues on this.

```
Content-Disposition: form-data; name="file"; filename="IMG_0001.HEIC"
Content-Type: image/jpeg

(blob)
-----------------------------ABCABCABCABCABCABCABCABCABCABC
Content-Disposition: form-data; name="message"

Hello World
-----------------------------ABCABCABCABCABCABCABCABCABCABC
etc.
```

**Comment by HoneyryderChuck** [comment]
Hi @forthrin ,  do check [the wiki about multipart support](https://gitlab.com/os85/httpx/-/wikis/Pass-Parameters#multipartform-data). But.long story short, yes.

**Comment by forthrin** [comment]
Ah. I didn't look for the right keywords. Almost there...

Current code:

```ruby
    form = {
      file: {
        content_type: 'image/jpeg',
        filename: 'foo.jpg',
        body: nil # for now
      },
      bar: {
        body: 'hello'
      },
      baz: {
        body: 'world'
      }
    }
```

Server expects (as verbatim as possible): 

```
-----------------------------*****************************
Content-Disposition: form-data; name="file"; filename="foo.jpg"
Content-Type: image/jpeg

-----------------------------*****************************
Content-Disposition: form-data; name="bar"

hello
-----------------------------*****************************
Content-Disposition: form-data; name="baz"

world
-----------------------------*****************************--
```

HTTPX generates:

```
-----------------------******************************************
Content-Disposition: form-data; name="file"; filename="foo.jpg"
Content-Type: image/jpeg


-----------------------******************************************
Content-Disposition: form-data; name="bar[body]"
Content-Type: text/plain

hello
-----------------------******************************************
Content-Disposition: form-data; name="baz[body]"
Content-Type: text/plain

world
-----------------------******************************************--
```

How does one:

1. Remove "[body]"
2. Remove "Content-Type: text/plain"
3. Set the separator value

**Reply by HoneyryderChuck** [reply]
It is recommended,  when working with files, to just pass the File handle to ":form". Otherwise,  I guess the question is answered?

**Comment by forthrin** [comment]
Possibly point 2 and 3 can be as-is, but I think "[body]" must away, otherwise the server won't identify the parts, containing vital metadata about the upload. Why is "[body]" added? Requirement by an RFC? Common convention? Other reason?

**Reply by HoneyryderChuck** [reply]
I see what you mean now.

1. It should already. That may be a bug. Can you create an issue? Feel free to investigate it yourself, as I'll only have ime next week.
2. Any reason why? Fwiw having it there or not has the same meaning (defaults to text/plain)
3. You cannot, and I don't think one needs to, unless you describe a common use case for it.

**Comment by forthrin** [comment]
https://github.com/HoneyryderChuck/httpx/blob/master/lib/httpx/transcoder.rb#L20

```Ruby
      elsif value.respond_to?(:to_hash)
        value.to_hash.each do |child_key, child_value|
          normalize_keys("#{key}[#{child_key}]", child_value, cond, &block)
        end
```

It seems to work like this. Setting "content_type" will probably work for my case, but I'd like to know the logic behind.)

- If omitting "content_type", square brackets are set for all elements, all elements are included.
- If setting "content_type", square brackets are omitted, and only the "body" element is included.

```Ruby
  form = {
    foo: {
      head: '1',
      body: '2'
    },
    bar: {
      content_type: 'text/plain',
      head: '3',
      body: '4',
      foot: '5',
    }
  }
  HTTPX.post(url, form:)

```

```
-----------------------3c150125dbd9cd1638a36d6be76f116ccb2d094945
Content-Disposition: form-data; name=foo[head]
Content-Type: text/plain

1
-----------------------3c150125dbd9cd1638a36d6be76f116ccb2d094945
Content-Disposition: form-data; name=foo[body]
Content-Type: text/plain

2
-----------------------3c150125dbd9cd1638a36d6be76f116ccb2d094945
Content-Disposition: form-data; name=bar
Content-Type: text/plain

4
-----------------------3c150125dbd9cd1638a36d6be76f116ccb2d094945--
```

**Reply by HoneyryderChuck** [reply]
Ok, that actually makes sense, now that I think about it. I.e. you either fill **all** required fields to turn it into a multipart component, or httpx will assume nesting. I don't think it makes sense to treat single use of ":body" differently.

I'll see if I can document this better though.
